L1Trigger/Phase2L1ParticleFlow/src/regionizer/middle_buffer_multififo_regionizer_ref.cpp

```diff
@@ -0,0 +1,632 @@
+      emCaloRegionizerPre_(nem, nem, false, outii, pauseii),
+      emCaloRegionizerPost_(nem, (nem + outii - 1) / outii, true, outii, pauseii),
+      muRegionizerPre_(nmu, nmu, false, outii, pauseii),
+      muRegionizerPost_(nmu, std::max(1u, (nmu + outii - 1) / outii), true, outii, pauseii),
+      tkBuffers_(ntk ? nbuffers_ : 0),
+      hadCaloBuffers_(ncalo ? nbuffers_ : 0),
+      emCaloBuffers_(nem ? nbuffers_ : 0),
+      muBuffers_(nmu ? nbuffers_ : 0) {
+  assert(nbuffers_ == nregions_post_ || nbuffers_ == nregions_pre_);
+  unsigned int phisectors = 9, etaslices = 3;
+  for (unsigned int ietaslice = 0; ietaslice < etaslices && ntk > 0; ++ietaslice) {
+    for (unsigned int ie = 0; ie < 2; ++ie) {  // 0 = negative, 1 = positive
+      unsigned int nTFEtaSlices = ietaslice == 1 ? 2 : 1;
+      if ((ietaslice == 0 && ie == 1) || (ietaslice == 2 && ie == 0))
+        continue;
+      unsigned int ireg0 = phisectors * ietaslice, il0 = 3 * NTK_LINKS * (nTFEtaSlices - 1) * ie;
+      for (unsigned int is = 0; is < NTK_SECTORS; ++is) {  // 9 tf sectors
+        for (unsigned int il = 0; il < NTK_LINKS; ++il) {  // max tracks per sector per clock
+          unsigned int isp = (is + 1) % NTK_SECTORS, ism = (is + NTK_SECTORS - 1) % NTK_SECTORS;
+          tkRoutes_.emplace_back(is + NTK_SECTORS * ie, il, is + ireg0, il0 + il);
+          tkRoutes_.emplace_back(is + NTK_SECTORS * ie, il, isp + ireg0, il0 + il + NTK_LINKS);
+          tkRoutes_.emplace_back(is + NTK_SECTORS * ie, il, ism + ireg0, il0 + il + 2 * NTK_LINKS);
+        }
+      }
+    }
+  }
+  // calo
+  for (unsigned int ie = 0; ie < etaslices; ++ie) {
+    for (unsigned int is = 0; is < NCALO_SECTORS; ++is) {  // NCALO_SECTORS sectors
+      for (unsigned int j = 0; j < 3; ++j) {               // 3 regions x sector
+        for (unsigned int il = 0; il < HCAL_LINKS; ++il) {
+          caloRoutes_.emplace_back(is, il, 3 * is + j + phisectors * ie, il);
+          if (j) {
+            caloRoutes_.emplace_back((is + 1) % 3, il, 3 * is + j + phisectors * ie, il + HCAL_LINKS);
+          }
+        }
+        for (unsigned int il = 0; il < ECAL_LINKS; ++il) {
+          emCaloRoutes_.emplace_back(is, il, 3 * is + j + phisectors * ie, il);
+          if (j) {
+            emCaloRoutes_.emplace_back((is + 1) % 3, il, 3 * is + j + phisectors * ie, il + ECAL_LINKS);
+          }
+        }
+      }
+    }
+  }
+  // mu
+  for (unsigned int il = 0; il < NMU_LINKS && nmu > 0; ++il) {
+    for (unsigned int j = 0; j < nregions_pre_; ++j) {
+      muRoutes_.emplace_back(0, il, j, il);
+    }
+  }
+}
+
+l1ct::MiddleBufferMultififoRegionizerEmulator::~MiddleBufferMultififoRegionizerEmulator() {}
+
+void l1ct::MiddleBufferMultififoRegionizerEmulator::initSectorsAndRegions(const RegionizerDecodedInputs& in,
+                                                                          const std::vector<PFInputRegion>& out) {
+  assert(!init_);
+  init_ = true;
+  assert(out.size() == nregions_post_);
```

Comment:
Since the amount of asserts here is relatively low, it would be probably better to throw `cms::Exception`s which can have some more information, than just an assert, which will only mention the line number of the failed assertion on failure.

Reply:
I can't use cms::Exception's since this code is also compiled outside of CMSSW in Vitis HLS testbenches.
I could use std exceptions, but eventually these asserts are not supposed to ever trigger when the code is running, they're there just to catch mistakes during development and to inform the reader of the assumptions that this piece of code is doing, so there's not much gain in making them more verbose.